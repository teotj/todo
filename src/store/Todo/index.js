const ADD_ITEM = 'ADD_ITEM';
const COMPLETE_ITEM = 'COMPLETE_ITEM';
const DELETE_ITEM = 'DELETE_ITEM';

const initialState = {
  items: []
};

// actions
export const addItem = (item) => ({ type: ADD_ITEM, item });
export const completeItem = (id) => ({ type: COMPLETE_ITEM, id });
export const deleteItem = (id) => ({ type: DELETE_ITEM, id });

// reducers
const addItemReducer = (state, action) => {
  const id = `-${Math.random().toString(36).substr(2, 9)}-`;
  const newItem = { title: action.item, isCompleted: false, id };

  return {
    ...state,
    items: [...state.items, newItem]
  };
};

const completeItemReducer = (state, action) => ({
  ...state,
  items: state.items.map((item) => ({ ...item, isCompleted: item.id === action.id || item.isCompleted }))
});

const deleteItemReducer = (state, action) => ({
  ...state,
  items: state.items.filter(({ id }) => id !== action.id)
});

const reducers = (state = initialState, action) => {
  switch (action.type) {
    case ADD_ITEM:
      return addItemReducer(state, action);
    case COMPLETE_ITEM:
      return completeItemReducer(state, action);
    case DELETE_ITEM:
      return deleteItemReducer(state, action);
    default:
      return state;
  }
};

export { reducers };
