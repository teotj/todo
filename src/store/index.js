import { createStore } from 'redux';

import { addItem, deleteItem, completeItem, reducers } from './Todo';

export const actions = { addItem, deleteItem, completeItem };
export const store = createStore(
  reducers,
  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__() // eslint-disable-line no-underscore-dangle
);
