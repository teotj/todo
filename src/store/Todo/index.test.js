import { addItem, deleteItem, completeItem, reducers } from 'store/Todo';

describe('Todo store actions', () => {
  const initialState = {
    items: [
      { title: 'titleA', id: '-idA-', isCompleted: false },
      { title: 'titleB', id: '-idB-', isCompleted: false }
    ]
  };

  it('Creates an action to add a todo item', () => {
    const expectedAction = { type: 'ADD_ITEM', item: 'todoA' };
    expect(addItem('todoA')).toEqual(expectedAction);
  });

  it('Creates an action to complete a todo item', () => {
    const expectedAction = { type: 'COMPLETE_ITEM', id: '-idA-' };
    expect(completeItem('-idA-')).toEqual(expectedAction);
  });

  it('Creates an action to delete a todo item', () => {
    const expectedAction = { type: 'DELETE_ITEM', id: '-idB-' };
    expect(deleteItem('-idB-')).toEqual(expectedAction);
  });

  it('Adds item when ADD_ITEM is called', () => {
    const mockAction = { type: 'ADD_ITEM', item: 'TodoA' };
    expect((reducers(undefined, mockAction)).items[0].title).toEqual('TodoA');
  });

  it('Completes item when COMPLETE_ITEM is called', () => {
    const mockAction = { type: 'COMPLETE_ITEM', id: '-idA-' };
    expect(reducers(initialState, mockAction)).toMatchSnapshot();
  });

  it('Deletes item when DELETE_ITEM is called', () => {
    const mockAction = { type: 'DELETE_ITEM', id: '-idA-' };
    expect(reducers(initialState, mockAction)).toEqual({
      items: [{ title: 'titleB', id: '-idB-', isCompleted: false }]
    });
  });

  it('Returns default state when unknown action is called', () => {
    const mockAction = { type: 'DEFAULT', id: '-idC-' };
    expect(reducers(initialState, mockAction)).toEqual(initialState);
  });
});
