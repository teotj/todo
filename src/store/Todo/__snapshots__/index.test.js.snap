// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo store actions Completes item when COMPLETE_ITEM is called 1`] = `
Object {
  "items": Array [
    Object {
      "id": "-idA-",
      "isCompleted": true,
      "title": "titleA",
    },
    Object {
      "id": "-idB-",
      "isCompleted": false,
      "title": "titleB",
    },
  ],
}
`;
